### DISCUSSION TITLE
Optional, non-nullable body parameters

### DISCUSSION THREAD

**Discussion by t-o-o-m** [fastapi/fastapi#12948] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class Hero(table=True):
    id: int
    name: str

class HeroUpdate:
    name: str | None = None

@app.patch("/heroes/{hero_id}")
def update_hero(hero_id: int, hero: HeroUpdate, session: SessionDep):
    hero_db = session.get(Hero, hero_id)
    hero_data = hero.model_dump(exclude_unset=True)
    hero_db.sqlmodel_update(hero_data)
    session.add(hero_db)
    session.commit()
    session.refresh(hero_db)
    return hero_db
```

### Description

Taking the shortened example from https://fastapi.tiangolo.com/tutorial/sql-databases/#herocreate-the-data-model-to-create-a-hero:

`name` of `HeroUpdate` is marked as `str | None` which is translated to a nullable, non-required property in the generated OpenAPI spec `(string | null)`
![grafik](https://github.com/user-attachments/assets/0b48672e-df6f-4aca-8f87-9c651c8e310a)

Following this spec, it is possible to send `null` values via `HeroUpdate.name`, but since `Hero.name` is actually not nullable this leads to uncaught exceptions if not further handled. 

Is there any possibility to define **non-required**, **non-nullable** fields? 

It seems that you can not define a non-required field without making it nullable (Well, you can. Setting `name: str = None` generates a correct OpenAPI spec but this will obviously fail type checks)

Especially for the example linked above this would be necessary; users should not be forced to send all parameters of a model for a patch-endpoint, and they also should not be allowed to send null values for parameters that are not nullable.

### Operating System

Linux

### FastAPI Version

0.114.2

### Pydantic Version

2.9.2

### Python Version

3.12.6

**Comment by luzzodev** [comment]
Hi @t-o-o-m i can confirm that documentation should be fixed because ,to be valid, name should be nullable on table too.

About your specific question:
```python
class HeroUpdate(BaseModel):
    name: str | None = None
```

If `name` is not passed, then `exclude_unset` work fine.
```
>> HeroUpdate().model_dump(exclude_unset=True)
>> {}
```

If `name` is passed as null , then `exclude_unset`  work fine but is not what you are expecting in your case
```

>> HeroUpdate(name=None).model_dump(exclude_unset=True)
>> {'name': None}
```

If `name` is passed as null , then `exclude_unset` and  `exclude_none` do what you are expecting in your case
```
>> HeroUpdate(name=None).model_dump(exclude_unset=True, exclude_none=True)
>> {}
```

Note:
You can use this combination of `exclude_unset` and  `exclude_none` if and only if `None` is not a valid value of your fields.

**Reply by t-o-o-m** [reply]
Thanks @luzzodev, I did not know about `exclude_none`! 
From a data-handling perspective, this is exactly what I am looking for. 

The OpenAPI specification still remains wrong, though, since the fields are listed as nullable even though they should be not. 
For now I'm using `name: str = None  # type: ignore`, which works perfectly fine, creates a correct OpenAPI spec and does not allow users to send `null` values, because the Pydantic validator kicks in before that:

```
{
  "detail": [
    {
      "type": "string_type",
      "loc": [
        "body",
        "name"
      ],
      "msg": "Input should be a valid string",
      "input": null
    }
  ]
}
```

I use the generated OpenAPI-spec to autogenerate clients, and I want this spec to be the single source of truth. Without the "hack" from above I'd need to explain to consumers that the values of type e.g. `str | None` are not _actually_ nullable, but that I can't define optional (=non-required) parameters otherwise.

**Reply by luzzodev** [reply, marked as answer]
If you define the `HeroUpdate` in this way

``` python
class HeroUpdate(BaseModel):
    name: str = Field(default=None, description="")
```

You will have several improvements:

- OpenAPI spec corrected.
- Right validation not allowing null values (default value are not validated)
- You can use `HeroUpdate().model_dump(exclude_unset=True)` without additional args to prevent dropping valid None values.

**Reply by t-o-o-m** [reply]
This is it! Thank you very much!
